base/task.jl

```diff
@@ -365,6 +365,60 @@ function wait(t::Task)
     nothing
 end
 
+# Wait multiple tasks
+waitany(tasks) = _wait_multiple(tasks)
+waitall(tasks; failfast=false) = _wait_multiple(tasks; all=true, failfast=failfast)
+
+function _wait_multiple(waiting_tasks::Union{AbstractVector{Task},AbstractSet{Task},Tuple{Task}}; all=false, failfast=false)::NTuple{2,Vector{Task}}
```

Comment:
I guess `typeassert` can be used to remove the restriction of the type of the first argument like below, but I don't know whether this is a general programming pattern in Julia.  I cannot find any other use cases to check the `eltype` with `typeassert` in this repository.

```suggestion
function _wait_multiple(waiting_tasks; all=false, failfast=false)::NTuple{2,Vector{Task}}
    typeassert(eltype(waiting_tasks), Type{Task})
```

Reply:
Yeah, I believe it's best to avoid type annotations for arguments here. While using `eltype` is an option, we could also consider iterating over `waiting_tasks` and directly applying `typeassert` to each element.
```suggestion
function _wait_multiple(waiting_tasks; all=false, failfast=false)::NTuple{2,Vector{Task}}
    foreach(waiting_tasks) do @nospecialize t
        if !(t isa Task)
            error("Expected an iterator of `Task` object")
        end
    end
```
And then we could also enhance type stability in the main `enumerator` for-loop by allowing you to specify `t = t::Task`.
```julia
    for (i, t) in enumerate(waiting_tasks)
        t = t::Task
        ...
```

Reply:
Thank you for your feedback.  I'd like to adopt your suggestion because it can accept inputs that consist only of tasks but have an `Any` eltype, such as `Vector{Any}`.  My version cannot accept such types of inputs.